src/dmd/e2ir.d

```diff
@@ -268,6 +268,43 @@ private elem *callfunc(const ref Loc loc,
                  */
                 ea.Ety = TYllong;
             }
+
+            /* Do integral promotions. This is not necessary per the C ABI, but
+             * some code from the C world seems to rely on it.
```

Comment:
Yes, it is necessary for the C ABI.  On x86_64, all integral types smaller than `int` should actually be passed as an `int`.  This apply to both parameters in the protoype and variadic parameters.

Reply:
My reading of the ABI doesn't make this clear. 16274 is the only case I've ever heard of where it's a problem. I also tries various cases with gcc, and on the receiving end I can't get it to rely on the upper bits.

Reply:
I was reading the compiler docs: https://gcc.gnu.org/onlinedocs/gccint/Stack-Arguments.html#TARGET_005fPROMOTE_005fPROTOTYPES

For x86_64 targets, this hook is set to true.